core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -512,16 +484,11 @@ class Partition(val topicPartition: TopicPartition,
       // record the epoch of the controller that made the leadership decision. This is useful while updating the isr
       // to maintain the decision maker controller's epoch in the zookeeper path
       controllerEpoch = partitionStateInfo.basePartitionState.controllerEpoch
-      // add replicas that are new
-      val newInSyncReplicas = partitionStateInfo.basePartitionState.isr.asScala.map(_.toInt)
-      newInSyncReplicas.filter(_ != localBrokerId).foreach(getOrCreateReplica)
-      inSyncReplicas = newInSyncReplicas.toSet
-
-      // remove assigned replicas that have been removed by the controller
-      val newAssignedReplicas = partitionStateInfo.basePartitionState.replicas.asScala.map(_.toInt)
-      (remoteReplicasMap.keys -- newAssignedReplicas).foreach(removeReplica)
-      newAssignedReplicas.filter(_ != localBrokerId).foreach(getOrCreateReplica)
 
+      updateAssignmentAndIsr(
+        assignment = partitionStateInfo.basePartitionState.replicas.asScala.iterator.map(_.toInt).toSeq,
```

Comment:
Same as above, since it's rather subtle, maybe we should mention that we convert the List from PartitionState to a Seq to preserve the order

Reply:
I added a comment to the `updateAssignmentAndIsr` method. Let me know if that address your concern.